Lib/inspect.py

```diff
@@ -2600,96 +2611,72 @@ def _signature_from_callable(obj, *,
+        # pure Python __new__ or __init__ method
+        for base in obj.__mro__:
+            # Now we check if the 'obj' class has an own '__new__' method
+            if new is not None and '__new__' in base.__dict__:
+                sig = _get_signature_of(new)
+                if skip_bound_arg:
+                    sig = _signature_bound_method(sig)
+                return sig
+            # or an own '__init__' method
+            elif init is not None and '__init__' in base.__dict__:
+                return _get_signature_of(init)
+
+        # At this point we know, that `obj` is a class, with no user-
+        # defined '__init__', '__new__', or class-level '__call__'
+
+        for base in obj.__mro__[:-1]:
+            # Since '__text_signature__' is implemented as a
+            # descriptor that extracts text signature from the
+            # class docstring, if 'obj' is derived from a builtin
+            # class, its own '__text_signature__' may be 'None'.
+            # Therefore, we go through the MRO (except the last
+            # class in there, which is 'object') to find the first
+            # class with non-empty text signature.
             try:
-                sig = _get_signature_of(call)
-            except ValueError as ex:
-                msg = 'no signature found for {!r}'.format(obj)
-                raise ValueError(msg) from ex
-
-    if sig is not None:
-        # For classes and objects we skip the first parameter of their
-        # __call__, __new__, or __init__ methods
-        if skip_bound_arg:
-            return _signature_bound_method(sig)
-        else:
-            return sig
+                text_sig = base.__text_signature__
+            except AttributeError:
+                pass
+            else:
+                if text_sig:
+                    # If 'base' class has a __text_signature__ attribute:
+                    # return a signature based on it
+                    return _signature_fromstr(sigcls, base, text_sig)
+
+        # No '__text_signature__' was found for the 'obj' class.
+        # Last option is to check if its '__init__' is
+        # object.__init__ or type.__init__.
+        if type not in obj.__mro__:
+            # We have a class (not metaclass), but no user-defined
+            # __init__ or __new__ for it
+            if (obj.__init__ is object.__init__ and
+                obj.__new__ is object.__new__):
+                # Return a signature of 'object' builtin.
+                return sigcls.from_callable(object)
+            else:
+                raise ValueError(
+                    'no signature found for builtin type {!r}'.format(obj))
 
-    if isinstance(obj, types.BuiltinFunctionType):
-        # Raise a nicer error message for builtins
-        msg = 'no signature found for builtin function {!r}'.format(obj)
-        raise ValueError(msg)
+    else:
+        # An object with __call__
+        call = getattr_static(type(obj), '__call__', None)
```

Comment:
So this can deal with non user defined methods? And that's why it's not using `_signature_get_user_defined_method(obj, '__call__')`? And that would be fixed in the future? I mean personally, I would just pass another (optional) argument to `_signature_get_user_defined_method` for whether to allow non user defined methods, but if this is going to be fixed soon, I'm okay for this being here temporarily.

Reply:
`_signature_get_user_defined_method` is more than 600 lines away from this place. You need to scroll 25 pages up and down to follow the logic.

Reply:
That's simply the case for almost all function calls. I don't think how far away the function is should be the deterministic factor for whether to use a function - otherwise we should just inline all the code. From my perspective, the "readability" of the code is how well the readers can understand the purpose of the code, and `_signature_get_user_defined_method` is explanatory.

So the question really is - is this trying to get the user defined method? Why it was different (not checking `_NonUserDefinedCallables`). Is it because the `_NonUserDefinedCallables` actually works in this case, or it's impossible to get `_NonUserDefinedCallables`(so it's actually okay to directly call `_signature_get_user_defined_method`)? 

I wouldn't have a clue why `getattr_static` is used or why we need to do a `_descriptor_get` without some deep knowledge - basically I would've lost here. But if it's a function `_signature_get_user_defined_method`, without knowing the detail of the implemenentation, I would at least understand - oh you are getting user defined method here, and that's often enough for reading the code here. That's why I wanted to put these logic into a function. In the future, if someone realize that a specific logic should be changed here, they would not need to search all the `getattr_static` usage, which probably scatters around, they only need to work in the only function.

Also, for any modern IDEs, 600 lines away is just a simple click - I don't think anyone is scrolling their screen to follow the function anymore. In VSCode, it's `Ctrl+click` to follow the function, `Alt+left arrow` to go back to the caller. Even with vim, that's a quick search in the same file.

Reply:
The _NonUserDefinedCallables actually works in this case, and the main purpose of this PR is to support signature when `__call__` is _NonUserDefinedCallables.

It actually mostly works in other three cases, but with caveats, different for different cases, so I prefer to leave this for other PR. This is why I plan to replace complex `_signature_get_user_defined_method` (its name and description will no longer be correct) and replace it with simpler 2-4 lines for every concrete case.

Reply:
Hmm okay, that makes sense. So if the function is not getting user defined method anymore (that it can take system-defined functions), we should at least change it. I guess we'll see if it's better to just remove the function and do a couple of lines for each usage, or to create another function with a explanatory name so the user would know that all these places are doing similar things.